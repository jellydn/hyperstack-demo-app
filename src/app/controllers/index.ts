// naming and imports: name and import as you see fit
import Auth from './auth'
import { NotesController } from './notes'

const controllers = []
controllers.push(Auth)
controllers.push(NotesController)

export const AppController = {
  controllers,
}
