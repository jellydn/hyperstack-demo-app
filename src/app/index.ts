import type { Context } from 'hyperstack'
import { context } from 'hyperstack'
import type * as Models from './models'
import type * as Mailers from './mailers'

export const appContext = context as Context<typeof Models, typeof Mailers>
