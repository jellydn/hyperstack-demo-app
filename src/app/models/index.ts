export { Note } from './note'
export { User } from './user'
