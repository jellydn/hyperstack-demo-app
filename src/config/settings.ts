// general purpose settings, with no infrastructure loading, no assumptions. keep it simple.
import path from 'path'
export const root = path.join(__dirname, '..')
