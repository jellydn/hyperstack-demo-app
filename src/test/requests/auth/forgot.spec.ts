import { test } from '@hyperstackjs/testing'
import { root } from '../../../config/settings'
import { appContext } from '../../../app'
const {
  requests,
  matchers: { matchRequestWithSnapshot, matchDeliveriesWithSnapshot },
  serializers: { redactAndExpectMatch, redactAndExpectMatchInEmails },
} = test(root)

const serializer = redactAndExpectMatch({
  'body.token': /^eyJh.+/,
})

const serializeTokenEmails = redactAndExpectMatchInEmails(
  /resetToken=[a-z0-9]{64}/g
)

const serializeUser = redactAndExpectMatch({
  pid: /.+-.+-.+-.+-.+/,
})

describe('requests', () => {
  describe('/auth/forgot', () => {
    requests(
      'should start forgot password flow and send email',
      async (request) => {
        const { User } = appContext.models()
        const user = await User.createWithPassword({
          username: 'evh@example.com',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        })
        await matchRequestWithSnapshot(
          200,
          request().post('/auth/forgot').send({
            username: 'evh@example.com',
          })
        )
        await user.reload()

        expect(user.resetToken).toMatch(/.{3,}/)
        expect(user.resetSentAt).toBeTruthy()
        expect(new Date().getTime() - user.resetSentAt!.getTime()).toBeWithin(
          0,
          30 * 1000
        )

        matchDeliveriesWithSnapshot(1, { serializer: serializeTokenEmails })
      }
    )
    requests(
      'should finish forgot password flow and reset password',
      async (request) => {
        const { User } = appContext.models()
        const user = await User.createWithPassword({
          username: 'evh@example.com',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        })
        await user.forgotPassword()

        const resetToken = user.resetToken

        await matchRequestWithSnapshot(
          400,
          request().post(`/auth/reset`).send({ password: 'short', resetToken }),
          { serializer, snapshotName: 'dont bypass our validation' }
        )

        await matchRequestWithSnapshot(
          400,
          request().post(`/auth/reset`).send({ resetToken: user.resetToken }),
          { serializer, snapshotName: 'you have to have a password' }
        )

        const newPassword = 'this-is-a-new-password'
        await matchRequestWithSnapshot(
          200,
          request()
            .post(`/auth/reset`)
            .send({ password: newPassword, resetToken }),
          { serializer, snapshotName: 'ok lets go' }
        )
        await user.reload()
        expect(user.resetToken).toBeNull()
        expect(user.resetSentAt).toBeNull()
        expect(await user.verifyPassword('mypass-should-login')).toBeFalse()
        expect(await user.verifyPassword('this-is-a-new-password')).toBeTrue()

        // and you cannot reuse this token
        await matchRequestWithSnapshot(
          400,
          request().post(`/auth/reset`).send({
            password: 'some-other-password',
            resetToken,
          }),
          { serializer, snapshotName: 'cannot reuse token' }
        )

        expect(serializeUser(user)).toMatchSnapshot()
        matchDeliveriesWithSnapshot(0)
      }
    )

    requests('should validate', async (request) => {
      const { User } = appContext.models()
      await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      await matchRequestWithSnapshot(
        401,
        request().post('/auth/forgot').send({
          username: 'no-such-user@example.com',
        }),
        { snapshotName: 'no such user' }
      )
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/forgot').send({
          username: 'this-is-not-email',
        }),
        { snapshotName: 'not an email' }
      )
      // no user
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/forgot').send({}),
        { snapshotName: 'no user field' }
      )

      // no token
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/reset').send({}),
        { snapshotName: 'no reset token field' }
      )

      // bad token
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/reset').send({
          resetToken: 'foobarbaz',
        }),
        { snapshotName: 'bad reset token' }
      )
    })
  })
})
