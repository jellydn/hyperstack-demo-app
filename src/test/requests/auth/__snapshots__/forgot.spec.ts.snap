// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests /auth/forgot should finish forgot password flow and reset password 1`] = `
{
  "name": "Eddie Van Halen",
  "pid": "test-redacted",
  "username": "evh@example.com",
}
`;

exports[`requests /auth/forgot should finish forgot password flow and reset password 2`] = `[]`;

exports[`requests /auth/forgot should finish forgot password flow and reset password: cannot reuse token 1`] = `
{
  "body": {
    "error": "missing, illegal, or expired reset token",
  },
  "headers": {
    "connection": "close",
    "content-length": "52",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should finish forgot password flow and reset password: dont bypass our validation 1`] = `
{
  "body": {
    "error": "password too short",
  },
  "headers": {
    "connection": "close",
    "content-length": "30",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should finish forgot password flow and reset password: ok lets go 1`] = `
{
  "body": {
    "ok": true,
  },
  "headers": {
    "connection": "close",
    "content-length": "11",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /auth/forgot should finish forgot password flow and reset password: you have to have a password 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "password",
        ],
        "received": "undefined",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "151",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should start forgot password flow and send email 1`] = `
{
  "body": {
    "ok": true,
  },
  "headers": {
    "connection": "close",
    "content-length": "11",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /auth/forgot should start forgot password flow and send email 2`] = `
[
  {
    "content": undefined,
    "contentType": "multipart/alternative",
    "contents": [
      undefined,
      "Reset your password with this link:

http://localhost/reset?test-redacted
",
      "<html>

<body>
  You can
  <a href="http://localhost:/reset?test-redacted">
    reset your password
  </a>
</body>

</html>
",
    ],
    "envelope": {
      "from": "no-reply@elle-postage.example.com",
      "to": [
        "evh@example.com",
      ],
    },
    "from": "no-reply@elle-postage.example.com",
    "messageId": "test-redacted",
    "originalMessage": {
      "attachments": [],
      "from": "Elle Postage <no-reply@elle-postage.example.com>",
      "html": "<html>

<body>
  You can
  <a href="http://localhost:/reset?test-redacted">
    reset your password
  </a>
</body>

</html>
",
      "subject": "Your reset password link",
      "text": "Reset your password with this link:

http://localhost/reset?test-redacted
",
      "to": "evh@example.com",
    },
    "response": "test-redacted",
    "subject": "Your reset password link",
    "to": [
      "evh@example.com",
    ],
  },
]
`;

exports[`requests /auth/forgot should validate: bad reset token 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "password",
        ],
        "received": "undefined",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "151",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should validate: no reset token field 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "resetToken",
        ],
        "received": "undefined",
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "password",
        ],
        "received": "undefined",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "261",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should validate: no such user 1`] = `
{
  "body": {
    "error": "incorrect username or password",
  },
  "headers": {
    "connection": "close",
    "content-length": "42",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`requests /auth/forgot should validate: no user field 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "username",
        ],
        "received": "undefined",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "151",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/forgot should validate: not an email 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_string",
        "message": "Invalid email",
        "path": [
          "username",
        ],
        "validation": "email",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "136",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;
